### ISSUE TITLE
Refactor PTYProcess to share code with t.i.process.Process

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7103]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spiv reported|
|-|-|
|Trac ID|trac#435|
|Type|enhancement|
|Created|2003-12-16 16:23:48Z|

Attachments:

* [process.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/784/784ef0059227d57909c9d81632b89915adc03c22/4eb421eb29f695708a54bee956d52909b4dfcffb.diff) (5791 bytes) - added by spiv on 2003-12-16 16:23:48Z - 
* [process-fdmap.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/784/784ef0059227d57909c9d81632b89915adc03c22/b50631b5a38f49fb45d85dad2e7c2405d66b2efe.diff) (52123 bytes) - added by warner on 2004-01-01 07:08:53Z - 
* [process-fdmap.diff2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/784/784ef0059227d57909c9d81632b89915adc03c22/6651dca6c84d5ef813ef02af1008109c9b90dd0e.diff2) (54612 bytes) - added by warner on 2004-01-04 14:47:47Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__435 435
type__enhancement enhancement
reporter__spiv spiv
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1071591828000000 1071591828000000
changetime__1469671985248094 1469671985248094
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__warner cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
As discussed on IRC, here's the beginnings of a patch to generalise FD handling
in process.Process to make it useful for inetd and gpg --status-fd.

This code still needs tests.  Also, warner wanted to make "in" and "out" file
descriptors be handled better, so I've assigned this to him.

Note this also fixes a buglet in the way fds are closed in the child.  It tries
to use resource.getrlimit(resource.RLIMIT_NOFILE)[1] rather than assuming the
maximum number of fds is 256.  This bit probably belongs in a new bug.

This patch also has a little bit of unrelated stuff in it: minor changes to
logging code.  Just ignore those bits if you don't like them :)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I haven't written up a sample implementation yet, but the API I'd like to see on
this would probably look like:

 childFDs={ 2: 0, 3: 1, 4: "r", 5: "rw", 6: "w" }

If the value is a number, it represents which of the parent's FDs should be dup'ed.

If the value is "r" or "rw", a new pipe will be connected to that child FD and
the ProcessProtocol's .outputReceived(fileno, data) method will be invoked when
data is read from that pipe. The 'fileno' argument is the child FD number. The
default ProcessProtocol implementation of .outputReceived() will call
self.outReceived or self.errReceived if fileno is 1 or 2, and will ignore the
data otherwise. This will retain backwards compatibility

If the value is "w" or "rw", a new pipe will be connected and the
.writeInput(fileno, data) method can be used to send data into the pipe. For
backwards compatibility, .transport.write will somehow turn into .writeInput(0,
data). I guess .transport.loseConnection probably still needs to work too.

The other API that needs improvement is pipe-closing. I think there should be a
dict that lets you check to see which pipes are still open, and a method that is
called each time a pipe is closed (due to EOF or the child closing the other
end). This should allow some flexibility in deciding which kinds of lost-pipes
should cause the child to be killed off.

 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
Oh, and of course there needs to be a method to close each pipe individually.
Probably PP.closePipe(fileno) . GPG uses this to know when you've finished
writing in the passphrase.
 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
Here's a larger patch which implements a modified API, adds a simple test case,
and provides some preliminary documentation.

The new API is:
 spawnProcess takes a 'childFDs' dict with keys that are either:
   a number, indicating a parent FD to dup
   "r", indicating the parent will read from the child
   "w", indicating the parent will write to the child
  The default argument is {0:"w", 1:"r", 2:"r"}.
 ProcessProtocol now has new methods:
  .dataReceived(self, childFD, data)
  .connectionLost(self, childFD)
 The protocol can now do more things to the child:
  pp.transport.writeToChild(childFD, data)
  pp.transport.closeChildFD(childFD)

I'm open to better suggestions on the method names. Note that everything is
expressed in terms of childFD numbers (the keys of the 'childFDs' dictionary
first passed into spawnProcess). The parent's end of the pipe used for each FD
is arbitrary and not useful as an index.

 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I'd rather it be called childDataReceived, and childConnectionLost, or something
like that. Are dup()ed fds registered with the event loop?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Itamar Shtull-Trauring &lt;itamar@...> added the comment:
> I'd rather it be called childDataReceived, and childConnectionLost, or something

Those sound fine to me.  inetd doesn't care about this part anyway ;)

> like that. Are dup()ed fds registered with the event loop?

The dup calls only happen in the child process, so the event loop is irrelevant.

In the parent process, see under the comment "arrange for the parent-side
pipes to be read and written:" for where startReading is called on the newly
created pipes.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
Here's a new patch, using childDataReceived and childConnectionLost . It also
includes the helper child program for the unit test, which was missing from the
previous patch.
 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
This has been committed.  The issue can be closed, yes?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
Probably. The remaining issues are:

 PTYProcess should be refactored to use the same code. (you can't use any
additional file descriptors with PTYs, you only get the in/out pair, but it
would be cleaner to have it use the separate ProcessReader and ProcessWriter design)

 It would be nice if Win32Process could do something similar.

 The new functionality fails on freebsd, one of the dup2 calls fails and I don't
know why.

 It could use an extra test case or two.

I'll add the extra test case and then close this one. I'll file a new bug on the
 freebsd failure (who's our freebsd wizard?), and will do the PTYProcess
refactoring next week.

If anyone has great ambitions to make Win32Reactor multiple-fd capable, go for
it. If that works, then my petmail stuff will be able to run on windows too,
which would be great.

 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Looks like this code is working on FreeBSD now? (We just got our FreeBSD
buildslave back - select reactor is passing all tests, at least).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
yeah. I'll leave this at done-cbb as a reminder that Win32Process and PTYProcess
still need to be refactored.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_12"></a>|@warner commented|
|-|-|

Nobody volunteered to help with the Win32Process side, so I'm renaming this ticket to just cover PTYProcess

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Isn't this refactoring what [#2341](https://github.com/twisted/twisted/issues/2341) is about?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

[#2341](https://github.com/twisted/twisted/issues/2341) is closed now.  Process and PTYProcess share a bunch of code.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|GitHub commented|
|-|-|

In changeset baba4c661a6606a56d8dab053cfe7b336a3c593e
```
#!CommitTicketReference repository="" revision="baba4c661a6606a56d8dab053cfe7b336a3c593e"
Merge pull request #435 from twisted/8683-ssh-channel-name-bytestring-1

Authord: einarfd
Reviewer: rodrigc
Fixes: #8683 

Convert ssh channel name to bytestring
```
